src/core/meshcop/secure_transport.cpp

```diff
@@ -1146,8 +1145,39 @@ void SecureTransport::Process(void)
 
     if (shouldDisconnect)
     {
-        Disconnect(event);
+        Disconnect(state);
+    }
+}
+
+void SecureTransport::NotifyConnectEvent(void)
```

Comment:
Can we consider moving this logic within the `SetState()`? I think this may work. Basically whever state changes, we can check if we need to signal any event?

Reply:
Moving this logic to `SetState()` changes the current logic. Currently for disconnect events there is a delay of `kGuardTimeNewConnectionMilli` between `SetState` and notification. Not quite sure if it would break anything if we notify disconnections earlier.

Reply:
You're right. We can't combine these. I'm not sure why the delay was originally introduced, but I wouldn't recommend changing it.